cgrok/tree.h

```diff
@@ -0,0 +1,44 @@
+#include <stdbool.h>
+#include <stdlib.h>
+#include "dict.h"
+
+typedef struct TCTREE TCTREE;
+
+// A replacement for the 32-bit int key comparator
+int tccmpint32(const void* k1, const void* k2);
+
+// The default comparator is for strings
+int dict_var_str_cmp(const void* k1, const void* k2);
+
+TCTREE *tctreenew(void);
+TCTREE *tctreenew2(dict_compare_func cp, void *cmpop);
+void tctreeiterinit(TCTREE *tree);
+const void *tctreeiternext(TCTREE *tree, int *sp);
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz);
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz);
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp);
+void tctreedel(TCTREE *tree);
+void tctreeclear(TCTREE *tree);
+
+// List functions
+
+typedef struct TCLISTELEM TCLISTELEM;
+
+typedef struct {
+  TCLISTELEM *head;
+  int len;
+} TCLIST;
+
+struct TCLISTELEM {
+  int size;
+  void *elem;
```

Comment:
nit: I'd call this `val` as `elem` confuses the data in the list with an element of the list. Or maybe the struct should be called `TCLISTNODE` as that's the terminology that I'm more used to.